[AST/Sema] RuntimeMetadata: Align implementation with proposal

xedin opened the pull request:
- @reflectionMetadata attributes declared on a protocol act as requirements. If type doesn't declare attribute, conformance fails unless conformance is stated directly on type declaration, in such cases attribute would be inferred.
- allow using reflection metadata attributes on unavailable extensions declared in the same module as type.
- Reject reflection metadata attributes with custom arguments on protocol declarations. Arguments cannot be used if types declare a conformance to a protocol from a different module.
- Diagnose situations where a sub-class or a protocol are missing reflection metadata attributes required by a superclass.

Resolves: rdar://103990788

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[AST] RuntimeMetadata: Allow use of reflection metadata attrs on unavailable extensions

This is just a way to opt-out a type from use of the attribute,
there is no code generation for such cases because the attribute
could never be reached.

xedin committed:
[AST] RuntimeMetadata: Reject attributes with custom arguments on protocol declarations

xedin committed:
[Sema] RuntimeMetadata: Make reflection metadata attribute a conformance requirement

If conforming type doesn't declare required set of reflection metadata
attributes (based on the protocol declaration) fail conformance via
extension.

LucianoPAlmeida commented on test/type/runtime_discoverable_attrs.swift:
```diff
@@ -264,3 +277,44 @@ extension EnumFlag where B == Void {
+}
+
+@available(*, unavailable)
+@EnumFlag extension EnumTypeTest { // expected-error {{@EnumFlag is already applied to type 'EnumTypeTest'; did you want to remove it?}} {{275:1-11=}}
```
There is probably a reason for it, but why the fix-it is removing the original instead of this occurrence of `@EnumFlag`?

xedin replied:
Because if you are trying to add unavailable extension then you don't really want this flag on the declaration.

xedin replied:
In other words - these declarations are mutually exclusive and based on the extension it's always better to remove attribute from the type, one might also consider keeping the one on extension in place though if type is a class to opt-out the type and all it's sub-classes from use of the attribute. (I still need to implement a diagnostic for sub-classes).

LucianoPAlmeida replied:
Ah I see, thanks for detailed explanation @xedin! I was more curious, because to me if one was trying to add an extension with attribute that already is on original decl, it doesn't need to be added on extension,so it would sounded reasonable remove from there, but I missed the availability bit.

xedin committed:
[Sema] RuntimeMetadata: Diagnose missing explicit reflection metadata attributes

Diagnose situations where a sub-class or a protocol do not have all
of the reflection metadata attributes required by a superclass.

Resolves: rdar://103990788

xedin committed:
[AST] RuntimeMetadata: Move `getRuntimeDiscoverableAttrs` to `Decl`

Reflection metadata attributes could be attached to a `ValueDecl`
and `ExtensionDecl`.

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test

xedin committed:
[Sema] RuntimeMetadata: Make sure that attrs on unavailable extensions always have a type

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test macOS platform
